0015-estimator-interface.md

```diff
@@ -0,0 +1,457 @@
+  1. Suppose that we want to give users the option of additionally returning the covariances between estimates that arise because of the circuit multiplexing, then we could update with the field `{"cov": <20,30,20,30>}`.
+  2. Suppose we want to return some indication of which estimates came from the same physical circuit.
+
+## Migration Path <a name="migration-path"></a>
+
+We need to remain backwards compatible with the existing interface to adhere to the [Qiskit Deprecation Policy](https://qiskit.org/documentation/deprecation_policy.html).
+
+In summary, we propose the following strategy: If the user has provided no `TaskLike`s, proceed with the old API, the old API output, and emit a deprecation warning, or an error if something mandatory like `observables` has been omitted. Otherwise, proceed with the new API, raising if they have tried to use the old arguments in addition to providing tasks.
+
+We now describe the strategy in detail, beginning with a restatement of the current interface, where `Estimator.run` has `circuits` as the only positional argument and accepts `observables` and `parameter_values` as keyword arguments (in addition to the var keyword `**run_options` arguments which isn't affected):
+
+```python
+class Estimator(BasePrimitive):
+
+    # current signature
+    def run(
+            self,
+            circuits: Sequence[QuantumCircuit] | QuantumCircuit,
+            observables: Sequence[BaseOperator | PauliSumOp | str] | BaseOperator | PauliSumOp | str,
+            parameter_values: Sequence[Sequence[float]] | Sequence[float] | float | None = None,
+            **run_options,
+        ):
+        ...
+```
+
+We propose the migration to using tasks in two phases.
+
+### Deprecation Phase 1
+
+In the first phase:
+* Introduce `tasks: Sequence[EstimatorTask] | EstimatorTask` as the only positional argument, and make `circuits` a keyword argument.
+* Coerce the arguments to account for the fact that the only existing positional argument is `circuit: Sequence[QuantumCircuit] | QuantumCircuit`.
+* Check that the user does not attempt to mix the old/new APIs.
+* If necessary, coerce the old-API arguments (now all keyword arguments) into `EstimatorTask`s, **raise deprecation warning**.
+* Always eventually run with `Estimator._run(tasks: Sequence[EstimatorTask], **run_options)`.
+
+Here is a mock implementation, with missing type annotations as above:
+```python
+class Estimator(BasePrimitive):
+
+    def run(
+        self, 
+        tasks: Sequence[EstimatorTask] | EstimatorTask, 
+        circuits = None,
+        observables = None, 
+        parameter_values = None, 
+        **run_options
+    ):
+        # Coerce into standard form to later allow for `tasks` to be the only non-kwarg.
+        if isinstance(tasks, QuantumCircuit) or all(isinstance(task, QuantumCircuit) for task in tasks):
+            if circuits is not None:
+                raise ValueError("Cannot mix old and new APIs")
+            circuits = tasks
```

Comment:
Related to Jessie's comment https://github.com/Qiskit/RFCs/pull/51/files#r1347953285.
If we put `tasks` as the first argument, we need to take care of all possibilities as follows. It might be too complicated.
```python
estimator = Estimator(...)
estimator.run([qc], [obs], [params])  # case 1
estimator.run([qc], [obs], parameter_values=[params])   # case 2
estimator.run([qc], observables=[obs], parameter_values=[params])   # case 3
estimator.run(circuits=[qc], observables=[obs], parameter_values=[params])   # case 4
```

The current code can handle case 3 and 4 correctly but fails with case 1 and 2.

As for case 1, we need to set arguments as follows.
```python
parameter_values = observables
observables = circuits
circuits = tasks
```

As for case 2, we need to set arguments as follows.
```python
observables = circuits
circuits = tasks
```

It would be a bit complicated to handle all cases properly. It would be nice to move `tasks` to the end as Jessie suggested https://github.com/Qiskit/RFCs/pull/51/files#r1347953285.

In this case, users always need to use keyword argument `tasks = ...` until we remove the legacy arguments.
There is an option to use keyword only argument for `tasks`. Then, the interface of `tasks` does not change before and after the removal of the legacy argument.
```python
# with legacy arguments
def run(self, circuits=None, observables=None, parameter_values=None, *, tasks=None, **run_options):

# after removal of legacy arguments
def run(self, *, tasks=None, **run_options):
```

If we would like to allow positional argument for `tasks`, it might be good to introduce versioning such as `EstimatorV2` as Jessie commented  https://github.com/Qiskit/RFCs/pull/51/files#r1347956614.

Reply:
> It would be a bit complicated to handle all cases properly.

Another option might be to modify the signature of `run()` to use [position-only arguments](https://peps.python.org/pep-0570/).  Something like:

```python
def run(self, arg1=None, /, observables=None, parameter_values=None, *, circuits=None, tasks=None, **run_options):
    if arg1 is not None:
        if not isinstance(arg1, Iterable):
            arg1 = [arg1]
        if all(isinstance(task, EstimatorTask) for task in arg1):
            tasks = arg1
        else:
            circuits = arg1
```
with some added handling of corner cases, of course.

On one hand, this adds some complexity to the `run()` method.  However, it should work with all cases @t-imamichi mentioned above, and if the goal is for people to eventually migrate to the `Task` interface, the added complexity of handling `arg1` is a temporary stepping stone which only needs to exist for as long as the deprecation period of the old API.

Reply:
Thanks, Jim. Your solution looks nice.